### ISSUE TITLE
Extra semicolon after cargo clippy --fix

### ISSUE THREAD

**Issue by tvallotton** [rust-lang/rust-clippy#8336]
### Summary

you can access a reproducible example in this branch: [clippy-fix-error](https://github.com/tvallotton/models/tree/clippy-fix-error). It seems that clippy forgot to remove a semicolon but I'm not entirely sure. I ran `clippy --fix` from the `models/models` directory. 

### Version

```text
rustc 1.58.1 (db9d1b20b 2022-01-20)
binary: rustc
commit-hash: db9d1b20bba1968c1ec1fc49616d4742c1725b4b
commit-date: 2022-01-20
host: aarch64-apple-darwin
release: 1.58.1
LLVM version: 13.0.0
```

### Error output

<details><summary>Backtrace</summary>
  <p>

  ```
warning: `models` (lib) generated 46 warnings
warning: failed to automatically apply fixes suggested by rustc to crate `models`

after fixes were automatically applied the compiler reported errors within these files:

  * models/src/orm/mod.rs
  * models/src/private/scheduler/driver/migration.rs

This likely indicates a bug in either rustc or cargo itself,
and we would appreciate a bug report! You're likely to see 
a number of compiler warnings after this message which cargo
attempted to fix but failed. If you could open an issue at
https://github.com/rust-lang/rust/issues
quoting the full output of this command we'd be very appreciative!
Note that you may be able to make some more progress in the near-term
fixing code with the `--broken-code` flag

The following errors were reported:
error[E0308]: mismatched types
   --> models/src/private/scheduler/driver/migration.rs:86:28
    |
86  |     pub fn commit(self) -> Result<Option<Report>> {
    |            ------          ^^^^^^^^^^^^^^^^^^^^^^ expected enum `std::result::Result`, found `()`
    |            |
    |            implicitly returns `()` as its body has no tail or `return` expression
...
104 |         };
    |          - help: consider removing this semicolon
    |
    = note:   expected enum `std::result::Result<std::option::Option<private::scheduler::driver::report::Report>, error::Error>`
            found unit type `()`

warning: unused import: `ConnectOptions`
  --> models/src/orm/mod.rs:12:5
   |
12 |     ConnectOptions,
   |     ^^^^^^^^^^^^^^
   |
   = note: `#[warn(unused_imports)]` on by default

error: aborting due to previous error; 1 warning emitted

For more information about this error, try `rustc --explain E0308`.
Original diagnostics will follow.
  ```

  </p>
</details>

**Comment by J-ZhengLi** [comment]
Here's a much simpler example in [playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=304896d4824a8bb395a50170eb71c5aa)

I think the problem here is that you don't need to add a semicolon after `if..else`, and `[needless_return]` does not take it into account.
